File: 2-concurrency-and-paralelism/node/src/benchmark.concurrent.js
import { generateData } from './generateData.js';
import { isPrime } from './isPrime.js';

const N = parseInt(process.env.N || '100000', 10);
const workers = parseInt(process.env.WORKERS || '4', 10);

const data = generateData(N);
const sliceSize = Math.ceil(N / workers);

async function run() {
  const start = Date.now();
  const tasks = [];
  for (let i = 0; i < workers; i++) {
    const startIdx = i * sliceSize;
    const endIdx = Math.min(startIdx + sliceSize, N);
    const slice = data.slice(startIdx, endIdx);
    tasks.push(Promise.resolve().then(() => {
      let c = 0;
      let s = 0;
      for (const v of slice) {
        if (isPrime(v)) {
          c++;
          s += v;
        }
      }
      return { count: c, sum: s };
    }));
  }
  const results = await Promise.all(tasks);
  let totalCount = 0;
  let totalSum = 0;
  for (const res of results) {
    totalCount += res.count;
    totalSum += res.sum;
  }
  const elapsedSec = (Date.now() - start) / 1000;
  console.log(JSON.stringify({
    mode: 'concurrent',
    N,
    workers,
    primes: totalCount,
    sum: totalSum,
    time: elapsedSec
  }));
}

run().catch((err) => {
  console.error(err);
});

File: 2025_12_10__19_00/go/src/benchmark_concurrent.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"runtime"
	"strconv"
	"sync"
	"time"

	"concurrency-benchmark-go/src/primes"
)

type ResultadoConc struct {
	Mode    string  `json:"mode"`
	N       int     `json:"N"`
	Workers int     `json:"workers"`
	Primes  int     `json:"primes"`
	Sum     int     `json:"sum"`
	Time    float64 `json:"time"`
}

func main() {
	nStr := os.Getenv("N")
	if nStr == "" {
		nStr = "100000"
	}
	N, err := strconv.Atoi(nStr)
	if err != nil {
		fmt.Fprintln(os.Stderr, "valor inválido para N")
		os.Exit(1)
	}
	workersStr := os.Getenv("WORKERS")
	if workersStr == "" {
		workersStr = "4"
	}
	workers, err := strconv.Atoi(workersStr)
	if err != nil || workers < 1 {
		workers = 4
	}

	// força uso de 1 CPU (concorrência sem paralelismo real)
	runtime.GOMAXPROCS(1)

	data := primes.GenerateData(N)
	sliceSize := (N + workers - 1) / workers

	var wg sync.WaitGroup
	wg.Add(workers)
	counts := make([]int, workers)
	sums := make([]int, workers)

	start := time.Now()
	for i := 0; i < workers; i++ {
		i := i
		go func() {
			defer wg.Done()
			startIdx := i * sliceSize
			endIdx := startIdx + sliceSize
			if endIdx > N {
				endIdx = N
			}
			c := 0
			s := 0
			for _, v := range data[startIdx:endIdx] {
				if primes.IsPrime(v) {
					c++
					s += v
				}
			}
			counts[i] = c
			sums[i] = s
		}()
	}
	wg.Wait()
	totalCount := 0
	totalSum := 0
	for i := 0; i < workers; i++ {
		totalCount += counts[i]
		totalSum += sums[i]
	}
	elapsed := time.Since(start).Seconds()
	res := ResultadoConc{
		Mode:    "concurrent-goroutines",
		N:       N,
		Workers: workers,
		Primes:  totalCount,
		Sum:     totalSum,
		Time:    elapsed,
	}
	enc := json.NewEncoder(os.Stdout)
	enc.SetEscapeHTML(false)
	if err := enc.Encode(res); err != nil {
		fmt.Fprintln(os.Stderr, err)
	}
}
